### PR TITLE
Replace NIOSSLError.unableToAllocateBoringSSLObject with fatalError

### LINKED ISSUE #173: deprecate & remove throw NIOSSLError.unableToAllocateBoringSSLObject; replace by crash
instead of `throw NIOSSLError.unableToAllocateBoringSSLObject`, we should (like Swift does for all other failing memory allocations) just crash instead.

### PR DESCRIPTION
**Pull request by agnosticdev**
Motivation:

Provides a resolution to issue-173.
deprecate & remove throw NIOSSLError.unableToAllocateBoringSSLObject;
replace by crash #173

Modifications:

Removed all NIOSSLError.unableToAllocateBoringSSLObject and replaced
with fatalError to force a crash.

Result:

- In critical points where execution should not continue a fatalError will be thrown instead of an NIOSSLError.unableToAllocateBoringSSLObject.
- fixes #173

### CODE REVIEW

File: Sources/NIOSSL/SSLPKCS12Bundle.swift [hunk 1]
```diff
@@ -64,7 +64,7 @@ public struct NIOSSLPKCS12Bundle {
             cert.map { CNIOBoringSSL_X509_free($0) }
             pkey.map { CNIOBoringSSL_EVP_PKEY_free($0) }
             caCerts.map { CNIOBoringSSL_sk_X509_pop_free($0, CNIOBoringSSL_X509_free) }
-            throw NIOSSLError.unableToAllocateBoringSSLObject
+            fatalError("Failed to obtain cert and pkey from a PKC12 file")
```
Comment (Lukasa) [inline Sources/NIOSSL/SSLPKCS12Bundle.swift]: We can delete the cleanup code above: if we're going to `fatalError`, we gain very little from freeing this memory first.

**Comment by weissi** [comment]
@agnosticdev this looks great, thank you! Mind deprecating the `unableToAllocateBoringSSLObject` enum case?

**Comment by weissi** [comment]
@agnosticdev as @Lukasa just pointed out: deprecating the enum case will lead to warnings in the `==` function but if you don't mind, just delete the manual implementation, the Swift compiler should be able to synthesise that now

**Comment by Lukasa** [comment]
See also #164.

**Comment by agnosticdev** [comment]
Thank you for the feedback.  Latest changes just committed.

File: Sources/NIOSSL/SSLErrors.swift [hunk 2]
```diff
@@ -52,6 +52,7 @@ public typealias NIOBoringSSLErrorStack = [BoringSSLInternalError]
 /// Errors that can be raised by NIO's BoringSSL wrapper.
 public enum NIOSSLError: Error {
     case writeDuringTLSShutdown
+    @available(*, deprecated, message: "unableToAllocateBoringSSLObject is no longer available.")
```
Comment (Lukasa) [inline Sources/NIOSSL/SSLErrors.swift]: Can we reword this to say "can no longer be thrown"?

Comment (agnosticdev) [reply Sources/NIOSSL/SSLErrors.swift]: Sounds good.  Would this be just "unableToAllocateBoringSSLObject can no longer be thrown," or just "can no longer be thrown?"

Comment (Lukasa) [reply Sources/NIOSSL/SSLErrors.swift]: Let's use the full language.

Also see the discussion in the main thread re: the equatable implementation.

**Comment by agnosticdev** [comment]
Changes made.  Let me know if this is what you had in mind for the removal of `==`.

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Comment by Lukasa** [comment]
@weissi We believe the API checker is wrong here, correct? While we removed the explicit function, that's an ABI breaker, not an API breaker.

**Comment by agnosticdev** [comment]
Just seen that too.
```
02:35:30 /* Removed Decls */
02:35:30 Func NIOSSLError.==(_:_:) has been removed
```
Let me know if there's anything I can do on my end.

**Comment by weissi** [comment]
@agnosticdev thanks very much for pointing this out. We believe it's a false positive: https://bugs.swift.org/browse/SR-12245
